src/cart/crt_hg.c

```diff
@@ -601,6 +601,30 @@ crt_provider_dec_cur_ctx_num(bool primary, int provider)
 	prov_data->cpg_ctx_num--;
 }
 
+void
+crt_provider_put_ctx_idx(bool primary, int provider, int idx)
+{
+	struct crt_prov_gdata *prov_data = crt_get_prov_gdata(primary, provider);
+
+	prov_data->cpg_used_idx[idx] = false;
+}
+
+int
+crt_provider_get_ctx_idx(bool primary, int provider)
+{
+	struct crt_prov_gdata	*prov_data = crt_get_prov_gdata(primary, provider);
+	int			i;
+
+	for (i = 0; i < CRT_SRV_CONTEXT_NUM; i++) {
```

Comment:
sorry if that seems naive:
Is that really an index required to be passed down to mercury or just any user defined id for the context?
do you need an array table here or just a context ID that is used in the context and here you just maintain how many context you use? 
i.e. can you just always return idx++; and all you maintain are how many active indexes there are at any given point (not like you had before where the id returned was the max index in use)?

Reply:
we use context index in other parts of the cart code to access static cache uri/hg_addr tables, so it needs to be between {0 and CRT_SRV_CONTEXT_NUM-1}. Similarly it is used for endpoint address generation (base_addr + ctx_idx). Eventually we might get rid of those static tables, but for now it would be a big change and some places would still need a linear index

Reply:
got it. thx